### PR TITLE
[SPARK-27252][SQL] Make current_date() independent from time zones

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

This makes the `CurrentDate` expression and `current_date` function independent from time zone settings. New result is number of days since epoch in `UTC` time zone. Previously, Spark shifted the current date (in `UTC` time zone) according the session time zone which violets definition of `DateType` - number of days since epoch (which is an absolute point in time, midnight of Jan 1 1970 in UTC time). 

The changes makes `CurrentDate` consistent to `CurrentTimestamp` which is independent from time zone too. 

## How was this patch tested?

The changes were tested by existing test suites like `DateExpressionsSuite`.

### CODE REVIEW

File: docs/sql-migration-guide-upgrade.md [hunk 1]
```diff
@@ -101,7 +101,9 @@ displayTitle: Spark SQL Upgrading Guide
 
   - In Spark version 2.4 and earlier, the `current_timestamp` function returns a timestamp with millisecond resolution only. Since Spark 3.0, the function can return the result with microsecond resolution if the underlying clock available on the system offers such resolution.
 
-  - In Spark version 2.4 abd earlier, when reading a Hive Serde table with Spark native data sources(parquet/orc), Spark will infer the actual file schema and update the table schema in metastore. Since Spark 3.0, Spark doesn't infer the schema anymore. This should not cause any problems to end users, but if it does, please set `spark.sql.hive.caseSensitiveInferenceMode` to `INFER_AND_SAVE`.
+  - In Spark version 2.4 and earlier, when reading a Hive Serde table with Spark native data sources(parquet/orc), Spark will infer the actual file schema and update the table schema in metastore. Since Spark 3.0, Spark doesn't infer the schema anymore. This should not cause any problems to end users, but if it does, please set `spark.sql.hive.caseSensitiveInferenceMode` to `INFER_AND_SAVE`.
```
Comment (MaxGekk) [inline docs/sql-migration-guide-upgrade.md]: I fixed the typo: `abd` -> `and`

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 2]
```diff
@@ -61,21 +61,15 @@ trait TimeZoneAwareExpression extends Expression {
 @ExpressionDescription(
   usage = "_FUNC_() - Returns the current date at the start of query evaluation.",
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Is it good to specify the time zone in this usage text?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Updated docs (and comments) of the expression and the current_date() function.

**Comment by MaxGekk** [comment]
@srowen @cloud-fan WDYT of the changes?

**Comment by srowen** [comment]
I see your point, in that a date by itself without a time shouldn't be relative to a time zone.`current_date` seems like it should reflect the date in the session time zone though. I think of it as the date portion of the current time, and time is relative to a time zone. Is this matching some other DB behavior?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 3]
```diff
@@ -58,20 +58,19 @@ object ReplaceExpressions extends Rule[LogicalPlan] {
  */
 object ComputeCurrentTime extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
-    val currentDates = mutable.Map.empty[String, Literal]
-    val timeExpr = CurrentTimestamp()
-    val timestamp = timeExpr.eval(EmptyRow).asInstanceOf[Long]
-    val currentTime = Literal.create(timestamp, timeExpr.dataType)
+    lazy val currentDate = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala]: the overhead of `lazy val` might be bigger than always computing the value. Can we add `lazy val` when we have benchmarks to prove it's necessary?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala]: If you have some concerns about performance, I could remove `lazy` or use the approach with `currentDates.getOrElseUpdate`.

**Comment by MaxGekk** [comment]
@cloud-fan Just in case, are you ok with the changes in general?

**Comment by cloud-fan** [comment]
According to the SQL standard, timestamp is timezone dependent while date is not. For `current_date`, it seems the timezone shifting is reasonable. E.g. when I run `current_date` in a database, I'd expect the result matches my local calendar. Can we check other mainstreaming databases?

**Comment by MaxGekk** [comment]
> timestamp is timezone dependent while **date is not**. For current_date, it seems the timezone **shifting is reasonable**. 

The **current_date()** returns a value of DateType which is time zone independent, and actually stores number of days since epoch (in UTC time zone). And Spark assumes that in many places. How can it be shifted by time zone offset?

> when I run current_date in a database, I'd expect the result matches my local calendar.

I would guess you speak about textual representation of current_date result? If so, conversion `DateType` -> `StringType` should care about time zone offset which what the `date_format()` function does.

**Comment by srowen** [comment]
Maybe I'm confused; `DateType` isn't logically time-zone-dependent but it's stored in a way that is (time since epoch)? OK this makes more sense. So if I later render this value according to my session time zone, it would already also 'correct' for the timezone?

If that's the case then we just have a bug, sure. After the fix, `current_date()` formatted for my current time zone would return what I expect today is? would be the current date from midnight to midnight in my timezone?

**Comment by srowen** [comment]
Your argument here is compelling: https://github.com/apache/spark/pull/24181#discussion_r269092290

So, basically we already convert from absolute time to date w.r.t. UTC elsewhere?

**Comment by MaxGekk** [comment]
> > what the date_format() function does.
> ... Your argument here is compelling: #24181 (comment)

Yah, I missed that `date_format()` has the same behavior as other date renders in Spark. Fortunately everything has been already consistent from this side :-)

**Comment by cloud-fan** [comment]
> I would guess you speak about textual representation of current_date result

Ah yes, then +1 for this PR

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala]: I'd expect to see
```
...
val currentDate = ...
...
val currentTime = ...
plan transformAllExpressions {
  case CurrentDate() => currentDate
  case CurrentTimestamp() => currentTime
}
```

**Comment by srowen** [comment]
@cloud-fan sounds like we're good with this; what do you think about your comment at https://github.com/apache/spark/pull/24185/files#r269162289 ?

**Comment by MaxGekk** [comment]
> @cloud-fan sounds like we're good with this; what do you think about your comment at https://github.com/apache/spark/pull/24185/files#r269162289 ?

Sorry, I forget about the comment. I am about to change this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala]: I removed lazy

**Review by srowen — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Just to make consequences of the changes clear:
1. `current_date()` will **always** produce current date as a value of `DateType` in the `UTC` time zone.
2. All available conversions of `DateType` to `StringType` is time zone independent in Spark at the moment. See the comment: https://github.com/apache/spark/pull/24181#discussion_r269092290 . Produced strings are **always** textual representation of dates in the UTC time zone.

To take the current date in local time zone (defined by `spark.sql.session.timeZone`), `DATE` column can be casted to `TIMESTAMP` column, and former one can be converted `STRING` column. This works because `TimestampType` to `StringType` is time zone dependent.  For example:
```sql
select date_format(cast(current_date as TIMESTAMP), 'yyyy-MM-dd')
```

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by srowen** [comment]
@MaxGekk @cloud-fan I'm seeing this failure in `master`; looks related?
```
ComputeCurrentTimeSuite:
- analyzer should replace current_timestamp with literals
- analyzer should replace current_date with literals *** FAILED ***
  17984 was greater than or equal to 17983, but 17984 was not less than or equal to 17983 (ComputeCurrentTimeSuite.scala:64)
```
If it's just the test that needs to be adjusted, maybe just a follow up to patch it, not a revert.

**Comment by MaxGekk** [comment]
@srowen I am fixing the test, and preparing a PR.

**Comment by MaxGekk** [comment]
Here is the PR https://github.com/apache/spark/pull/24240 which fixes `ComputeCurrentTimeSuite`.

**Pull request closed without merging**